elastic/elastic-agent
# Implementing `RetryUntil` and `RetryableError` fields in Upgrade Details metadata

## What does this PR do?

This PR adds two new fields to Upgrade Details metadata: 
* `RetryUntil`: The deadline until when a retryable upgrade step, e.g. the download step, will be (re)tried.
* `RetryableError`: Any error that is a result of a retryable upgrade step, e.g. the download step, being (re)tried.

## Why is it important?

To know why a retryable upgrade step is being retried and for how long it will be retried.

## Checklist


## How to test this PR locally

1. Build Agent from this PR's branch
2. Edit `elastic-agent.yml` so the download URI is deliberately broken:
   ```yaml
   agent.download.sourceURI: https://artifacts.elastic.co/broken/
3. Install Agent
4. Attempt an upgrade.
5. In a new window, check the Agent status, human output. 
   ```
   sudo elastic-agent status --output human
   
   ┌─ fleet
   │  └─ status: (STOPPED) Not enrolled into Fleet
   ├─ elastic-agent
   │  └─ status: (UPGRADING) Upgrading to version 8.12.0-SNAPSHOT
   └─ upgrade_details
      ├─ target_version: 8.12.0-SNAPSHOT
   ├─ state: UPG_DOWNLOADING
   └─ metadata
      ├─ download_percent: 0.00%
      ├─ retry_until: 1h59m47.994034s
      └─ retryable_error_msg: unable to download package: 3 errors occurred:
             * package '/Library/Elastic/Agent/data/elastic-agent-8a8abd/downloads/elastic-agent-8.12.0-SNAPSHOT-darwin-aarch64.tar.gz' not found: open /Library/Elastic/Agent/data/elastic-agent-8a8abd/downloads/elastic-agent-8.12.0-SNAPSHOT-darwin-aarch64.tar.gz: no such file or directory
             * call to 'https://artifacts.elastic.co/broken/beats/elastic-agent/elastic-agent-8.12.0-SNAPSHOT-darwin-aarch64.tar.gz' returned unsuccessful status code: 404
             * call to 'https://artifacts.elastic.co/broken/beats/elastic-agent/elastic-agent-8.12.0-SNAPSHOT-darwin-aarch64.tar.gz' returned unsuccessful status code: 404
   ```
6. Verify that `upgrade_details.state` is set to `UPG_DOWNLOADING`.
7. Verify that `upgrade_details.metadata.retry_until` and `upgrade_details.metadata.retryable_error` are set.  
8. Verify that `upgrade_details.metadata.retry_until` shows the duration left for retries.

## Related issues

- Closes #3818
